### PR TITLE
[FIXED JENKINS-38473] Set a short timeout on the socket when using it to wake the acceptor thread.

### PR DESCRIPTION
**Pull request by stephenc**
[JENKINS-38473](https://issues.jenkins-ci.org/browse/JENKINS-38473)

@jenkinsci/code-reviewers @reviewbybees

### CODE REVIEW

File: core/src/main/java/hudson/TcpSlaveAgentListener.java [hunk 1]
```diff
@@ -258,6 +259,7 @@ public void run() {
          */
         private void respondHello(String header, Socket s) throws IOException {
             try {
+                s.
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/TcpSlaveAgentListener.java]: WAT?

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
PR is incomplete. Would be also great to make the timeout configurable via system properties since 1000 MAY be too small in edge cases

**Comment by oleg-nenashev** [comment]
🐛  for now. I think bot code leftover and timeout magic number should be addressed

File: core/src/main/java/hudson/TcpSlaveAgentListener.java (hunk 1 above)
Comment (stephenc) [inline core/src/main/java/hudson/TcpSlaveAgentListener.java]: Oh dear... how did that end up there... I'll fix on monday

**Comment by stephenc** [comment]
@oleg-nenashev if it takes longer than a second to wake the acceptor thread that doesn't matter as the event for wake will be in the works and we are just waiting for the acceptor thread to wake... could run a shorter timeout too but that risks more log statements about a socket being closed before any data was read.

In any case what advice would we give users about tuning this? Reality is none.

**Comment by jtnord** [comment]
I'm kind of :bee: here - but at the same time with the later JNLPv4 protocol we do not do blocking in the slecector thread - so this whole shutdown - send something to the JNLPPort to wake the selector should be uneeded...   Perhaps I'm missing something :-(

**Comment by oleg-nenashev** [comment]
> In any case what advice would we give users about tuning this? Reality is none.

Well, some users like thinking they manage something :)
Anyway, I'm more or less fine with this interval. 🐝

**Comment by oleg-nenashev** [comment]
@reviewbybees done

**Comment by stephenc** [comment]
@jtnord the thread that accepts incoming connections is blocking IO not NIO, even for JNLP4 if we switched to NIO we'd have to use a separate selector to ensure it works on windows.

The point of the ping is to wake the blocking IO thread by establishing an incoming connection.

File: core/src/main/java/hudson/TcpSlaveAgentListener.java [hunk 2]
```diff
@@ -170,6 +170,7 @@ public void shutdown() {
             if (localAddress instanceof InetSocketAddress) {
                 InetSocketAddress address = (InetSocketAddress) localAddress;
                 Socket client = new Socket(address.getHostName(), address.getPort());
+                client.setSoTimeout(1000); // waking the acceptor loop should be quick
```
Comment (jtnord) [inline core/src/main/java/hudson/TcpSlaveAgentListener.java]: Why not just set the SO_TIMEOUT on the server socket and remove whole section?

**Comment by jtnord** [comment]
maybe there is a reason I am not aware of - but the code looks overly complex.
There is no need to wake the selector thread if you use a socket timeout on the server socket.. and then there is no need to send a ping and then there is no need to check if any exception was because we where shutting down, and then the code that handles the accept can also control the shutting down of the socket - and then you have much cleaner code with less hacks...

**Comment by oleg-nenashev** [comment]
So, somebody can follow-up with a better fix later.
@jtnord Could you create a follow-up task?

**Review by oleg-nenashev — APPROVED** [review]

**Pull request merged**